api4/system.go

```diff
@@ -0,0 +1,908 @@
+	}
+}
+
+func databaseRecycle(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !c.App.SessionHasPermissionTo(*c.AppContext.Session(), model.PermissionRecycleDatabaseConnections) {
+		c.SetPermissionError(model.PermissionRecycleDatabaseConnections)
+		return
+	}
+
+	auditRec := c.MakeAuditRecord("databaseRecycle", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+
+	if *c.App.Config().ExperimentalSettings.RestrictSystemAdmin {
+		c.Err = model.NewAppError("databaseRecycle", "api.restricted_system_admin", nil, "", http.StatusForbidden)
+		return
+	}
+
+	c.App.RecycleDatabaseConnection()
+
+	auditRec.Success()
+	ReturnStatusOK(w)
+}
+
+func invalidateCaches(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !c.App.SessionHasPermissionTo(*c.AppContext.Session(), model.PermissionInvalidateCaches) {
+		c.SetPermissionError(model.PermissionInvalidateCaches)
+		return
+	}
+
+	auditRec := c.MakeAuditRecord("invalidateCaches", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+
+	if *c.App.Config().ExperimentalSettings.RestrictSystemAdmin {
+		c.Err = model.NewAppError("invalidateCaches", "api.restricted_system_admin", nil, "", http.StatusForbidden)
+		return
+	}
+
+	err := c.App.Srv().InvalidateAllCaches()
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	auditRec.Success()
+
+	w.Header().Set("Cache-Control", "no-cache, no-store, must-revalidate")
+	ReturnStatusOK(w)
+}
+
+func getLogs(c *Context, w http.ResponseWriter, r *http.Request) {
+	auditRec := c.MakeAuditRecord("getLogs", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+
+	if *c.App.Config().ExperimentalSettings.RestrictSystemAdmin {
+		c.Err = model.NewAppError("getLogs", "api.restricted_system_admin", nil, "", http.StatusForbidden)
+		return
+	}
+
+	if !c.App.SessionHasPermissionTo(*c.AppContext.Session(), model.PermissionGetLogs) {
+		c.SetPermissionError(model.PermissionGetLogs)
+		return
+	}
+
+	lines, err := c.App.GetLogs(c.Params.Page, c.Params.LogsPerPage)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	auditRec.AddMeta("page", c.Params.Page)
+	auditRec.AddMeta("logs_per_page", c.Params.LogsPerPage)
+
+	w.Write([]byte(model.ArrayToJSON(lines)))
+}
+
+func postLog(c *Context, w http.ResponseWriter, r *http.Request) {
+	forceToDebug := false
+
+	if !*c.App.Config().ServiceSettings.EnableDeveloper {
+		if c.AppContext.Session().UserId == "" {
+			c.Err = model.NewAppError("postLog", "api.context.permissions.app_error", nil, "", http.StatusForbidden)
+			return
+		}
+
+		if !c.App.SessionHasPermissionTo(*c.AppContext.Session(), model.PermissionManageSystem) {
+			forceToDebug = true
+		}
+	}
+
+	m := model.MapFromJSON(r.Body)
+	lvl := m["level"]
+	msg := m["message"]
+
+	if len(msg) > 400 {
+		msg = msg[0:399]
+	}
+
+	msg = "Client Logs API Endpoint Message: " + msg
+	fields := []mlog.Field{
+		mlog.String("type", "client_message"),
+		mlog.String("user_agent", c.AppContext.UserAgent()),
+	}
+
+	if !forceToDebug && lvl == "ERROR" {
+		mlog.Error(msg, fields...)
```

Comment:
Gentle reminder to check our logging [principles](https://developers.mattermost.com/contribute/server/style-guide/#log-levels) before merging this change.